Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -491,13 +496,40 @@ extension ActivityStreamPanel: DataObserverDelegate {
         }
     }
 
-    func getHighlights() -> Success {
+    func getBookmarksForHighlights() -> Deferred<Maybe<Cursor<Site>>> {
         let count = ASPanelUX.BookmarkHighlights // Fetch 2 bookmarks
-        return self.profile.recommendations.getHighlights().both(self.profile.recommendations.getRecentBookmarks(count)).bindQueue(.main) { (highlights, bookmarks) in
-            guard let highlights = highlights.successValue?.asArray(), let bookmarks = bookmarks.successValue?.asArray() else {
+        return self.profile.recommendations.getRecentBookmarks(count)
+    }
+
+    // Used to check if the entire section is turned off
+    // when it is we shouldnt show the emtpy state
+    func isHighlightsEnabled() -> Bool {
+        let bookmarks = profile.prefs.boolForKey(PrefsKeys.ASBookmarkHighlightsVisible) ?? false
+        let history = profile.prefs.boolForKey(PrefsKeys.ASRecentHighlightsVisible) ?? false
+        return history && bookmarks
+    }
+
+    func getHighlights() -> Success {
+        var queries: [() -> Deferred<Maybe<Cursor<Site>>>] = []
+        if profile.prefs.boolForKey(PrefsKeys.ASBookmarkHighlightsVisible) ?? true {
+            queries.append(getBookmarksForHighlights)
+        }
+
+        if profile.prefs.boolForKey(PrefsKeys.ASRecentHighlightsVisible) ?? true {
+            queries.append(self.profile.recommendations.getHighlights)
+        }
+
+        guard !queries.isEmpty else {
+            self.highlights = []
+            return succeed()
+        }
+
+        return accumulate(queries).bindQueue(.main) { result in
+            guard let resultArr = result.successValue else {
                 return succeed()
```

Comment:
Do we actually want to `succeed()` here?

Reply:
I'm using bindQueue here to chain a deferred. So accumulate returns a deferred which instead of returning that deferred to the caller it returns the deferred inside the block. Confusing.
But here is the method signature of bindQueue
`@escaping (T) -> Deferred.Deferred<U>`
vs 
uponQueue
` @escaping (T) -> ()`